File: src/main/java/ini/cx3d/cells/CellFactory.java
/*
Copyright (C) 2009 Frédéric Zubler, Rodney J. Douglas,
Dennis Göhlsdorf, Toby Weston, Andreas Hauri, Roman Bauer,
Sabina Pfister & Adrian M. Whatley.

This file is part of CX3D.

CX3D is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CX3D is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CX3D.  If not, see <http://www.gnu.org/licenses/>.
*/

package ini.cx3d.cells;

import ini.cx3d.localBiology.SomaElement;
import ini.cx3d.physics.PhysicalNode;
import ini.cx3d.physics.PhysicalSphere;
import ini.cx3d.simulations.ECM;
import ini.cx3d.spatialOrganization.SpatialOrganizationNode;
import ini.cx3d.utilities.Matrix;

import java.util.Vector;

/**
 * <code>CellFacory</code> generates a new  <code>Cell</code>, <code>SomaElement</code>, 
 * <code>PhysicalSphere</code> and <code>SpatialOrganizationNode</code>.  We set than the
 * massLocation and cell color.
 * We can generate a single <code>Cell</code> or a list of <code>Cell</code> distributed
 * uniformly.
 * @author rjd/sabina
 *
 */
public class CellFactory {
    
    /* Reference to the ECM. */
    private static ECM ecm = ECM.getInstance();

	/**
	 * <code>CellFactory</code> constructor.
	 */
    public CellFactory() { 
    }
    
    /**
     * Generates a single cell at the specified position.
     * @param cellOrigin
     * @return
     */
    public static Cell getCellInstance(double[] cellOrigin) {
    	
    	// Create new cell
        Cell cell = new Cell();
        SomaElement soma = new SomaElement();
        cell.setSomaElement(soma);        
        PhysicalSphere ps = new PhysicalSphere(); 
        soma.setPhysical(ps);
        SpatialOrganizationNode<PhysicalNode> son = ecm.getSpatialOrganizationNodeInstance(cellOrigin.clone(), ps);
        ps.setSoNode(son);
        
        // Add cell to ECM instance
        ecm.addPhysicalSphere(soma.getPhysicalSphere());
        
        // Set cell properties
        ps.setMassLocation(cellOrigin);
        ps.setColor(ecm.cellTypeColor(cell.getType()));
        return cell;
    }
    
   /**
    * Generates a 2D grid of cells according according to the desired number of cells along 
    * the y and x axes. Cell position can be randomized by increasing the standard deviation of
    * the Gaussian noise distribution. 
    * @param xmin 
    * @param xmax
    * @param ymin
    * @param ymax
    * @param nx: Number of cells along the x axis
    * @param ny: Number of cells along the y axis
    * @param noiseStd: Gaussian noise standard deviation
    * @return cellList
    */
   public static Vector<Cell> get2DCellGrid(double xmin, double xmax, double ymin,
		   double ymax, double zpos, int nx, int ny, double noiseStd) {
	   
	   // Insert all generated cells in a vector
       Vector<Cell> cellList = new Vector<Cell>();
       double dx = (xmax-xmin)/(1+nx);
       double dy = (ymax-ymin)/(1+ny);
       
       // Generate cells
       for (int i=1; i < nx+1; i++) {
       	for (int j=1; j < ny+1; j++) {
       		double[] newLocation = {
       				xmin+i*dx+ecm.getGaussianDouble(0, noiseStd), 
       				ymin+j*dy+ecm.getGaussianDouble(0, noiseStd), 
       				zpos + ecm.getGaussianDouble(0, noiseStd)};
       		Matrix.print(newLocation);
       		Cell cell = getCellInstance(newLocation);
       		cellList.add(cell);
       	}
       }
       return cellList;
   }
   
   /**
    * Generates a 3D grid of cells according to the desired number of cells along 
    * the y, x and z axes. Cell position can be randomized by increasing the standard deviation of
    * the Gaussian noise distribution. 
    * @param xmin 
    * @param xmax
    * @param ymin
    * @param ymax
    * @param zmin
    * @param zmax
    * @param nx: Number of cells along the x axis
    * @param ny: Number of cells along the y axis
    * @param nz: Number of cells along the z axis
    * @param noiseStd: Gaussian noise standard deviation
    * @return cellList
    */
   public static Vector<Cell> get3DCellGrid(double xmin, double xmax, double ymin,
		   double ymax, double zmin, double zmax, int nx, int ny, int nz, double noiseXYStd, double noiseZStd) {
	   
	   // Insert all generated cells in a vector
       Vector<Cell> cellList = new Vector<Cell>();
       double dx = (xmax-xmin)/(1+nx);
       double dy = (ymax-ymin)/(1+ny);
       double dz = (zmax-zmin)/(1+nz);
       
       // Generate cells
       for (int i=1; i < nx+1; i++) {
       	for (int j=1; j < ny+1; j++) {
       		for (int k=1; k < nz+1; k++) {
       			double[] newLocation = {
       					xmin+i*dx+ecm.getGaussianDouble(0, noiseXYStd), 
       					ymin+j*dy+ecm.getGaussianDouble(0, noiseXYStd), 
       					zmin+k*dz+ecm.getGaussianDouble(0, noiseZStd)};
       			Cell cell = getCellInstance(newLocation);
       			cellList.add(cell);
       		}
       	}
       }
       return cellList;
   }


}

File: src/cells/cell_factory.cc
#include "cells/cell_factory.h"

#include "physics/physical_node.h"
#include "physics/physical_sphere.h"
#include "local_biology/soma_element.h"
#include "spatial_organization/space_node.h"

namespace cx3d {
namespace cells {

using std::vector;

Cell* CellFactory::getCellInstance(const std::array<double, 3>& cell_origin,
                                   const std::shared_ptr<simulation::ECM>& ecm) {
  // Create new cell
  auto cell = new Cell();
  auto soma = SomaElement::UPtr(new SomaElement());
  auto ps = PhysicalSphere::UPtr(new PhysicalSphere());
  auto son = ecm->getSpatialOrganizationNodeInstance(cell_origin, ps.get());
  ps->setSoNode(son);

  // Add cell to ECM instance
  ecm->addPhysicalSphere(ps.get());  //fixme critical

  // Set cell properties
  ps->setMassLocation(cell_origin);
  ps->setColor(ecm->cellTypeColor(cell->getType()));

  soma->setPhysical(std::move(ps));
  cell->setSomaElement(std::move(soma));
  return cell;
}

vector<Cell*> CellFactory::get2DCellGrid(double x_min, double x_max, double y_min, double y_max,
                                                         double z_pos, int n_x, int n_y, double noise_std,
                                                         const std::shared_ptr<simulation::ECM>& ecm) {

  // Insert all generated cells in a vector
  vector<Cell* > cell_list;
  double dx = (x_max - x_min) / (1 + n_x);
  double dy = (y_max - y_min) / (1 + n_y);

  // Generate cells
  for (int i = 1; i < n_x + 1; i++) {
    for (int j = 1; j < n_y + 1; j++) {
      std::array<double, 3> new_location { x_min + i * dx + ecm->getGaussianDouble(0, noise_std), y_min + j * dy
          + ecm->getGaussianDouble(0, noise_std), z_pos + ecm->getGaussianDouble(0, noise_std) };
      auto cell = getCellInstance(new_location, ecm);
      cell_list.push_back(cell);
    }
  }
  return cell_list;
}

vector<Cell*> CellFactory::get3DCellGrid(double x_min, double x_max, double y_min, double y_max,
                                                         double z_min, double z_max, int n_x, int n_y, int n_z,
                                                         double noise_xy_std, double noise_z_std,
                                                         const std::shared_ptr<simulation::ECM>& ecm) {

  // Insert all generated cells in a vector
  vector<Cell*> cell_list;
  double dx = (x_max - x_min) / (1 + n_x);
  double dy = (y_max - y_min) / (1 + n_y);
  double dz = (z_max - z_min) / (1 + n_z);

  // Generate cells
  for (int i = 1; i < n_x + 1; i++) {
    for (int j = 1; j < n_y + 1; j++) {
      for (int k = 1; k < n_z + 1; k++) {
        std::array<double, 3> new_location { x_min + i * dx + ecm->getGaussianDouble(0, noise_xy_std), y_min + j * dy
            + ecm->getGaussianDouble(0, noise_xy_std), z_min + k * dz + ecm->getGaussianDouble(0, noise_z_std) };
        auto cell = getCellInstance(new_location, ecm);
        cell_list.push_back(cell);
      }
    }
  }
  return cell_list;
}

}  // namespace cells
}  // namespace cx3d
